src/coreclr/jit/lower.cpp

```diff
@@ -7802,6 +7802,313 @@ void Lowering::ContainCheckBitCast(GenTree* node)
     }
 }
 
+struct StoreCoalescingData
+{
+    var_types targetType;
+    GenTree*  baseAddr;
+    GenTree*  index;
+    GenTree*  value;
+    uint32_t  scale;
+    int       offset;
+};
+
+//------------------------------------------------------------------------
+// GetStoreCoalescingData: given a STOREIND node, get the data needed to perform
+//    store coalescing including pointer to the previous node.
+//
+// Arguments:
+//    ind      - the STOREIND node
+//    data     - [OUT] the data needed for store coalescing
+//    prevTree - [OUT] the pointer to the previous node
+//
+// Return Value:
+//    true if the data was successfully retrieved, false otherwise.
+//    Basically, false means that we definitely can't do store coalescing.
+//
+static bool GetStoreCoalescingData(GenTreeStoreInd* ind, StoreCoalescingData* data, GenTree** prevTree)
+{
+    // Don't merge volatile stores.
+    if (ind->IsVolatile())
+    {
+        return false;
+    }
+
+    // Data has to be INT_CNS, can be also VEC_CNS in future.
+    if (!ind->Data()->IsCnsIntOrI())
+    {
+        return false;
+    }
+
+    // We will need to walk previous trees (up to 4 nodes in a store)
+    // We already know that we have value and addr nodes, LEA might add more.
+    const int MaxTrees        = 4;
+    GenTree*  trees[MaxTrees] = {ind->Data(), ind->Addr(), nullptr, nullptr};
+    int       treesCount      = 2;
+
+    data->targetType = ind->TypeGet();
+    data->value      = ind->Data();
+    if (ind->Addr()->OperIs(GT_LEA))
+    {
+        GenTree* base  = ind->Addr()->AsAddrMode()->Base();
+        GenTree* index = ind->Addr()->AsAddrMode()->Index();
+        if ((base == nullptr) || !base->OperIs(GT_LCL_VAR))
+        {
+            // Base must be a local. It's possible for it to be nullptr when index is not null,
+            // but let's ignore such cases.
+            return false;
+        }
+        if ((index != nullptr) && !index->OperIs(GT_LCL_VAR))
+        {
+            // Index is either nullptr or a local.
+            return false;
+        }
+        data->baseAddr = base == nullptr ? nullptr : base;
+        data->index    = index == nullptr ? nullptr : index;
+        data->scale    = ind->Addr()->AsAddrMode()->GetScale();
+        data->offset   = ind->Addr()->AsAddrMode()->Offset();
+
+        // Add index and base to the list of trees.
+        trees[treesCount++] = base;
+        if (index != nullptr)
+        {
+            trees[treesCount++] = index;
+        }
+    }
+    else if (ind->Addr()->OperIs(GT_LCL_VAR))
+    {
+        // Address is just a local, no offset, scale is 1
+        data->baseAddr = ind->Addr();
+        data->index    = nullptr;
+        data->scale    = 1;
+        data->offset   = 0;
+    }
```

Comment:
These presumably need checks for address exposure, otherwise the first `STOREIND` could modify one of the locals used in a later one.

Reply:
Thanks! Addressed, luckily it seems it didn't affect the diffs